lightning/src/ln/channelmanager.rs

```diff
@@ -6207,34 +6314,55 @@ where
 		}
 	}
 
-	fn peer_connected(&self, counterparty_node_id: &PublicKey, init_msg: &msgs::Init) -> Result<(), ()> {
+	fn peer_connected(&self, counterparty_node_id: &PublicKey, init_msg: &msgs::Init, inbound: bool) -> Result<(), ()> {
 		if !init_msg.features.supports_static_remote_key() {
 			log_debug!(self.logger, "Peer {} does not support static remote key, disconnecting with no_connection_possible", log_pubkey!(counterparty_node_id));
 			return Err(());
 		}
 
-		log_debug!(self.logger, "Generating channel_reestablish events for {}", log_pubkey!(counterparty_node_id));
-
 		let _persistence_guard = PersistenceNotifierGuard::notify_on_drop(&self.total_consistency_lock, &self.persistence_notifier);
 
+		// If we have too many peers connected which don't have funded channels, disconnect the
+		// peer immediately (as long as it doesn't have funded channels). If we have a bunch of
+		// unfunded channels taking up space in memory for disconnected peers, we still let new
+		// peers connect, but we'll reject new channels from them.
+		let connected_peers_without_funded_channels = self.peers_without_funded_channels(|node| node.is_connected);
+		let inbound_peer_limited = inbound && connected_peers_without_funded_channels >= MAX_NO_CHANNEL_PEERS;
```

Comment:
if we do not have any disconnection criteria, does this present a potential risk of resource exhaustion even without any adversaries in picture?

Reply:
IIUC, 
Currently if 250 inactive peers connect to me, nobody else can connect to me(even if they wanted to open a channel in next few seconds), as we drop their connections immediately.

Even after #1987, doesn't this introduce an additional DoS risk? (Maybe even bigger than the one we are fixing here, since it doesn't even need an adversary)

After 1987, how big do we expect these limits to be, so that it will be a non-issue?
Imo, without any disconnection criteria, the way we are immediately dropping connection from new inbound peer is always going to be a risk.

Reply:
in another comment you said "as nodes that are likely to open a channel are exempted anyway." 
-> dont we disconnect new inbound peer_connection immediately without giving them a chance ?

Reply:
> Even after https://github.com/lightningdevkit/rust-lightning/issues/1987, doesn't this introduce an additional DoS risk? (Maybe even bigger than the one we are fixing here, since it doesn't even need an adversary)

See the discussion starting here https://github.com/lightningdevkit/rust-lightning/pull/1988#pullrequestreview-1273397164. This PR constitutes a net improvement as we're moving away from a performance DoS restraining some critical operations of LDK to a denial-of-service in the sense of restraining operational continuity of the node (i.e not accepting new peer beyond 250 peers). A LDK infrastructure operator should be free to deploy additional physical nodes under the same logical pubkey (by shipping the same `NodeSigner` implem in multiple `ChannelManager`) and a frontend peer connection load-balancer can be deployed before our `PeerManager` (what Eclair does). We can still explore more exhaustion mitigation solutions independently of #1988.

Reply:
Sounds good, so we will eventually allowlist nodes with on-chain channel anywhere. 👍